contracts-bedrock/contracts/L1/ResourceMetering.sol

```diff
@@ -0,0 +1,188 @@
+     */
+    function _updateResourceParams() internal {
+        (uint128 prevBaseFee, uint64 prevBoughtGas, uint64 prevBlockNum) = getResourceParams();
+
+        uint256 blockDiff = block.number - prevBlockNum;
+        if (blockDiff == 0) {
+            return;
+        }
+
+        // Handle updating EIP-1559 style gas parameters. We use EIP-1559 to restrict the rate at
+        // which deposits can be created and therefore limit the potential for deposits to DoS the
+        // L2 system. Fee scheme is very similar to EIP-1559 with minor changes.
+        int128 gasTarget = int128(RESOURCE_LIMIT / ELASTICITY_MULTIPLIER);
+        int128 gasUsedDelta = int128(uint128(prevBoughtGas)) - gasTarget;
+        int256 baseFeeDelta = (int128(prevBaseFee) * gasUsedDelta) /
+            gasTarget /
+            int128(BASE_FEE_MAX_CHANGE_DENOMINATOR);
+
+        uint128 newBaseFee = _clamp(
+            int256(uint256(prevBaseFee)) + baseFeeDelta,
+            BASE_FEE_MIN,
+            type(uint128).max
+        );
+
+        // If we skipped more than one block, we also need to account for every empty block. Empty
+        // block means there was no demand for deposits in that block, so we should reflect this
+        // lack of demand in the fee.
+        if (blockDiff > 1) {
+            uint256 n = blockDiff - 1;
+            newBaseFee = _clamp(
+                int256((newBaseFee * (7**n)) / (8**n)),
+                BASE_FEE_MIN,
+                type(uint128).max
+            );
+        }
+
+        // Update resource parameters.
+        resources = SlotPacking128x64x64.set(newBaseFee, 0, uint64(block.number));
+    }
+
+    /**
+     * Uses a specified amount of the resource.
+     *
+     * @param _amount Amount of the resource to use.
+     * @param _gasCostStipend Amount of L1 gas used to get to this point (refunded).
+     */
+    function _useResources(uint64 _amount, uint256 _gasCostStipend) internal {
+        (uint128 prevBaseFee, uint64 prevBoughtGas, uint64 prevBlockNum) = getResourceParams();
+
+        // Make sure we can actually buy the gas that the user wants.
+        prevBoughtGas += _amount;
+        require(
+            prevBoughtGas <= RESOURCE_LIMIT,
+            "OptimismPortal: cannot buy more gas than available gas limit"
+        );
+
+        // Burn gas to pay for the deposit, based on the base fee and gas limit.
+        uint256 resourceCost = _amount * prevBaseFee;
+        uint256 gasCost = resourceCost / Math.max(block.basefee, INITIAL_BASE_FEE);
+        if (gasCost > _gasCostStipend) {
+            _burnGas(gasCost - _gasCostStipend);
+        }
+
+        // Update resource parameters.
+        resources = SlotPacking128x64x64.set(prevBaseFee, prevBoughtGas, prevBlockNum);
+    }
+
+    /**
+     * Clamps a value between two bounds.
+     *
+     * @param _value Value to clamp.
+     * @param _min Minimum bound.
+     * @param _max Maximum bound.
+     * @return Clamped value.
+     */
```

Comment:
See this https://github.com/foundry-rs/forge-std/commit/fe53c1391122211ec8b835abb637aed8e98c12ae

Reply:
Nit: I think that `_clamp` should be in a library